Update L2NormToNorm transformation

achetver opened the pull request:
Description: Update L2NormToNorm transformation and rename it to L2NormFusing - since IE doesn't support normalization with reduction along batch dimension, we cannot fuse related subgraph to NormalizeL2 operation in such cases.

Ticket: 44952

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names


Validation:
* [x]  Unit tests: N/A - No new implemented functions;
* [x]  Framework operation tests: N/A - No new enabled operations;
* [x]  Transformation tests: Updated;
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py: N/A - No affected changes;
* [x]  Model Optimizer IR Reader check: N/A - No affected changes.

Documentation:
* [x]  Supported frameworks operations list: N/A - No new supported operations;
* [x]  Supported **public** models list: N/A - Nothing to add;
* [x]  New operations specification: N/A - No new supported operations;
* [x]  Guide on how to convert the **public** model: N/A - Nothing to add;
* [x]  User guide update: N/A - Nothing to update.

achetver commented:
@iimironov @rkazants @vgavrilo this PR is ready for review. please, take a look.

rkazants commented on model-optimizer/extensions/middle/L2NormToNorm.py:
```diff
@@ -82,6 +82,17 @@ def replace_pattern(self, graph: Graph, match: dict):
+            axes = int64_array([axes])
+        axes.sort()
+        if not (np.array_equal(axes, int64_array([1, 2])) or np.array_equal(axes, int64_array([1, 2, 3]))):
+            log.debug('IE doesn\'t support l2 normalization with reduction along axes {}, skip fusing transformation.'
+                      ''.format(axes))
```
why do you use `log.debug`?

rkazants commented on model-optimizer/extensions/middle/L2NormToNorm.py:
```diff
@@ -82,6 +82,17 @@ def replace_pattern(self, graph: Graph, match: dict):
             log.debug('The value of the "maximum_y_data" is not defined or is not constant')
             return
 
+        # We need to check axes which performed reduction because IE supports reduction only along HW and HWC axes.
```
what about 5D rank tensors. May it be better to say: "... except batch dimension".

achetver replied:
From file `inference-engine/src/legacy_api/src/ngraph_ops/normalize_ie.cpp`:

`NODE_VALIDATION_CHECK(this, input_shape.rank().is_dynamic() || (input_shape.rank().get_length() >= 2 && input_shape.rank().get_length() <= 4), "Argument must have rank >= 2 and <= 4 (argument shape: ", input_shape, ").");`

Looks like now NormalizeIE operation doesn't support 5D tensors. I will update this commentary with this information.

achetver replied:
Comment was updated.

rkazants commented on model-optimizer/extensions/middle/L2NormToNorm.py:
```diff
@@ -82,6 +82,17 @@ def replace_pattern(self, graph: Graph, match: dict):
+        if axes.shape == ():
+            axes = int64_array([axes])
+        axes.sort()
+        if not (np.array_equal(axes, int64_array([1, 2])) or np.array_equal(axes, int64_array([1, 2, 3]))):
```
to me the most obvious check: np.any(axes == 0)

achetver replied:
This check was refactored and updated to be more clear and common.

achetver commented:
This transformation was refactored and updated to be more clear and common. Unit tests was refactored and added more test cases. Also these changes was successfully tested using layer tests and some models. @iimironov @rkazants @vgavrilo, please, take a look.

vgavrilo commented on model-optimizer/extensions/middle/L2NormToNorm.py:
```diff
@@ -91,7 +112,7 @@ def replace_pattern(self, graph: Graph, match: dict):
                                                           {'name': output_name, 'eps': y,
-                                                           'across_spatial': 0, 'channel_shared': 0})
+                                                           'across_spatial': across_spatial, 'channel_shared': False})
         rename_node(normalize_node, output_name)
```
Why all reduction axes are `1`? Why don't you specify the required attribute `eps_mode`? According the pattern, this attribute should be equal to `max`.

vgavrilo replied:
If you want to specify only axes `1`, maybe you should write `int64_array([1])`?

achetver replied:
This value was leave "as is", because it doesn't used in related transformation `extensions/back/NormalizeToNormalizeL2.py`. The same for `eps_mode` attribute. We decide to make this quick fix firstly, and after that make some changes in IE and related changes in MO.

vgavrilo commented on model-optimizer/extensions/middle/L2NormToNorm.py:
```diff
@@ -82,6 +82,27 @@ def replace_pattern(self, graph: Graph, match: dict):
 
+        # We need to check axes which performed reduction because IE supports only 2D, 3D, 4D inputs and
+        # reduction only along spatial dimensions.
+        input_shape = match['sum'].in_port(0).data.get_shape().size
```
Maybe `input_rank`, not `input_shape`?

achetver replied:
Yes, it is better name. I will change this name if there will be another change requests in this PR.

Feel free to reopen this discussion if you think what this change is necessary.

rkazants commented on model-optimizer/extensions/middle/L2NormToNorm.py:
```diff
@@ -82,6 +82,27 @@ def replace_pattern(self, graph: Graph, match: dict):
 
+        # We need to check axes which performed reduction because IE supports only 2D, 3D, 4D inputs and
+        # reduction only along spatial dimensions.
+        input_shape = match['sum'].in_port(0).data.get_shape().size
```
please rename this variable to `input_rank`. I got confusion.

achetver replied:
Done.

rkazants commented on model-optimizer/extensions/middle/L2NormToNorm.py:
```diff
@@ -82,6 +82,27 @@ def replace_pattern(self, graph: Graph, match: dict):
+        axes.sort()
+
+        across_spatial = False
+        if np.array_equal(axes, int64_array(np.arange(start=1, stop=input_shape))):
```
`1` is not spatial dimension.

achetver replied:
It is middle stage and this transformation is specific for TensorFlow only, so this operation in `NHWC` layout.

rkazants replied:
Please make a comment about this. Then you need to check layout and I think the correct reference operand `int64_array(np.arange(start=1, stop=input_shape-1))` where you also need to exclude `C` channel.

lazarevevgeny replied:
>It is middle stage and this transformation is specific for TensorFlow only, so this operation in NHWC layout.

TensorFlow models can be in NCHW layout. So the layout should be explicitly checked.

achetver replied:
across_spatial is a flag that specifies if normalization is performed over CHW or HW. Maybe be better if I rename this variable, but I am still think that logic is right.

rkazants replied:
what is purpose of  `channel_shared` ?

achetver replied:
`channel_shared` is a flag that specifies if scale parameters are shared across channels.
It is Caffe-specific attribute and looks like it doesn't applicable to this fused operation.

rkazants replied:
I got it about `channel_shared`, thanks. Regarding `across_spatial` I think you need to make a comment that it means normalization excepting batch dimension only. Otherwise, it creates confusion since channel dimension is not a spatial dimension.

achetver replied:
Transformation was refactored and this code was removed. In new implementation comments were updated.

rkazants commented on model-optimizer/extensions/middle/L2NormToNorm.py:
```diff
@@ -82,6 +82,27 @@ def replace_pattern(self, graph: Graph, match: dict):
+        if axes.shape == ():
+            axes = int64_array([axes])
+        axes.sort()
+
```
I think that `channel_shared` should be set up based `axes` as well as `across_spatial`. Now you ignore channel dimension.

lazarevevgeny commented on model-optimizer/extensions/middle/L2NormToNorm.py:
```diff
@@ -82,6 +82,27 @@ def replace_pattern(self, graph: Graph, match: dict):
             log.debug('The value of the "maximum_y_data" is not defined or is not constant')
             return
 
+        # We need to check axes which performed reduction because IE supports only 2D, 3D, 4D inputs and
```
I suggest to completely refactor this transformation and generate operation `NormalizeL2`. Because currently we generate non-opset operation "Normalize" which is then converted to `NormalizeL2` in the `model-optimizer/extensions/back/NormalizeToNormalizeL2.py`. Moreover, the `NormalizeL2` op semantic is exactly what we need: it has `axes` input , and `eps_mode`. So there is no need to bother with `across_channels` and `across_spatial` at all :)

achetver replied:
Ok, done.

lazarevevgeny reviewed (changes requested):
Suggestion is to rewrite the transformation completely and do not generate `Normalize` operation at all.

achetver commented:
@rkazants @iimironov @lazarevevgeny Transformation was refactored, now all checks have passed, please, review these changes.

rkazants commented on model-optimizer/extensions/middle/L2NormFusing.py:
```diff
@@ -82,16 +82,31 @@ def replace_pattern(self, graph: Graph, match: dict):
             log.debug('The value of the "maximum_y_data" is not defined or is not constant')
             return
```
I think that it is better to call this transformation as `FusionToNormalizeL2`

rkazants replied:
btw, do we really need this transformation in MO? Why don't we make it by nGraph?

lazarevevgeny replied:
Decided to not apply the comment.
